common/src/main/java/earth/terrarium/adastra/api/events/AdAstraEvents.java

```diff
@@ -43,6 +47,26 @@ static boolean post(ServerLevel level, LivingEntity entity) {
         }
     }
 
+    @FunctionalInterface
+    public interface EntityHasOxygenEvent {
+        Optional<Boolean> tick(Entity entity, boolean hasOxygen);
```

Comment:
I don't think you should use Optionals for the oxygen and gravity event. Every entity checks for oxygen and gravity, so creating a new Optional object every time it's checked would not be efficient. I'd suggest using something like `TriState` from resourcefullib. It's just an enum with 3 values: `TRUE` `FALSE` and `UNDEFINED` so you can mimic the same behavior without instantiating any new objects.

Reply:
Then in gravity, should I use `float`  and compare with a specific constant like -1?
Like this code.
```java
@FunctionalInterface
public interface EntityGravityEvent {
	public static final float NOT_HANDLED = Float.NaN;
	
	float getGravity(Entity entity, float gravity);

	static void register(EntityGravityEvent listener) {
		ENTITY_GRAVITY_LISTENERS.add(listener);
	}

	@ApiStatus.Internal
	static float post(Entity entity, float gravity) {
		for (var listener : ENTITY_GRAVITY_LISTENERS) {
			float newGravity = listener.getGravity(entity, gravity);
			if (newGravity != NOT_HANDLED) {
				return newGravity;
			}
		}
		return gravity;
	}
}
```

Reply:
Alternatively, there is another way to exit post when value returned from event listener is different with original value.

Reply:
Alternative solution makes the most sense. 👍

Reply:
Changed.